experimental: Add Maglev support

DamianSawicki opened the pull request:
This adds support for the Maglev load-balancing algorithm to the experimental StateDB load balancing.

**Update:** In addition to the original goal of the PR, which is contained in the first commit, @joamaki added additional commit(s) improving the existing Maglev implementation (support for concurrency, migration to hive cells, and better data structures to avoid rewriting data from one to another).

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

joamaki commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -430,6 +438,7 @@ func (ops *BPFOps) Prune(_ context.Context, _ statedb.ReadTxn, _ iter.Seq2[*Fron
 		ops.pruneBackendMaps(),
 		ops.pruneRevNat(),
 		ops.pruneSourceRanges(),
+		// TODO: Should there be any pruning for Maglev? If so, probably BPFOps should have additional fields, on which pruning would be based.
```
Yeah should add pruning for maglev as well. Should be easy enough: dump the outer map and delete each outer value that points to a non-existing frontend ID. Almost identical to `pruneRevNat`.

What do you mean by additional fields? I don't see why this needs any extra state.

DamianSawicki replied:
> What do you mean by additional fields? I don't see why this needs any extra state.

Other pruning methods rely on `ops.backendReferences` or `ops.backendStates`, so I was envisioning that something similar may be needed, but your previous paragraphs clarified that it's not. Thanks!

joamaki commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -601,6 +610,31 @@ func (ops *BPFOps) updateFrontend(fe *Frontend) error {
 
+	// Update Maglev
+	if ops.cfg.NodePortAlg == option.NodePortAlgMaglev {
+		ops.log.Info("Update Maglev", "feID", feID)
```
Should be `Debug`, but let's do an explicit separate pass to change all the `Info`s to `Debug`s.

joamaki commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -601,6 +610,31 @@ func (ops *BPFOps) updateFrontend(fe *Frontend) error {
 	}
 
+	// Update Maglev
+	if ops.cfg.NodePortAlg == option.NodePortAlgMaglev {
```
Perhaps move this into a helper as it's fair bit of lines and this method is getting big?

DamianSawicki replied:
Done.

joamaki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -90,6 +92,17 @@ type sourceRangeMaps interface {
 
+type maglevMaps interface {
+	/*
+		TODO: Perhaps it would be better to split UpdateMaglev into the following two methods:
```
Why would this be better? It seems much worse to me.

DamianSawicki replied:
The current setup places quite some logic into `UpdateMaglev()`:
 * similar logic is not present in analogous methods for the remaining interfaces, so I thought it is inconsistent,
 * the logic is duplicated in `*BPFLBMaps` and `*FakeLBMaps`, which means that testing with fakes does not validate the logic in the actual thing.

If we decide to check the inner map for existence before recreating, that would be even more logic inside `UpdateMaglev()`. I think I've only been testing with fakes in my IDE, which is why I was leaning towards moving as much logic from the maps as possible, but I'm also fine with keeping the current interface if you prefer.

joamaki replied:
I would prefer keeping this logic behind `LBMaps` and it's fine if FakeLBMaps duplicates some of it.

DamianSawicki replied:
Sure, comment removed.

joamaki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -443,6 +486,85 @@ func (r *BPFLBMaps) UpdateSourceRange(key lbmap.SourceRangeKey, value *lbmap.Sou
+// UpdateMaglev implements lbmaps.
+func (r *BPFLBMaps) UpdateMaglev(key lbmap.MaglevOuterKey, backendIDs []loadbalancer.BackendID, ipv6 bool) error {
+	/*
+		TODO: Should we delete the previous inner map (if it exists) from the kernel? Will it be garbage-collected automatically?
```
No need to delete it. The inner map is just a small array, so cheap to just re-create rather than update in place.

DamianSawicki replied:
Ok, thanks!

joamaki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -443,6 +486,85 @@ func (r *BPFLBMaps) UpdateSourceRange(key lbmap.SourceRangeKey, value *lbmap.Sou
+// UpdateMaglev implements lbmaps.
+func (r *BPFLBMaps) UpdateMaglev(key lbmap.MaglevOuterKey, backendIDs []loadbalancer.BackendID, ipv6 bool) error {
+	/*
+		TODO: Should we delete the previous inner map (if it exists) from the kernel? Will it be garbage-collected automatically?
```
No need to delete it. The inner map is just a small array, so cheap to just re-create rather than update in place.

DamianSawicki replied:
Duplicated comment, closing.

joamaki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -443,6 +486,85 @@ func (r *BPFLBMaps) UpdateSourceRange(key lbmap.SourceRangeKey, value *lbmap.Sou
+// UpdateMaglev implements lbmaps.
+func (r *BPFLBMaps) UpdateMaglev(key lbmap.MaglevOuterKey, backendIDs []loadbalancer.BackendID, ipv6 bool) error {
+	/*
+		TODO: Should we delete the previous inner map (if it exists) from the kernel? Will it be garbage-collected automatically?
```
Maps are refcounted, so no need to worry about deleting.

The inner map is a fixed-sized array. It's cheap to recreate. But yeah, we could probably just do a lookup and update the inner map directly. Maybe worth benchmarking if it's faster by any meaningful amount?

DamianSawicki replied:
Added a TODO, but marked it as optional in the light of your [other comment](https://github.com/cilium/cilium/pull/35430#discussion_r1808699110).

joamaki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -443,6 +486,85 @@ func (r *BPFLBMaps) UpdateSourceRange(key lbmap.SourceRangeKey, value *lbmap.Sou
+		return nil
+	}
+	/*
+		TODO: Should we use val to delete the inner table from the kernel or is it garbage-collected automatically?
```
BPF maps are reference counted and cannot be explicitly deleted. A reference can exist from an open file descriptor, from a pin file (in /sys/fs/bpf), or from another BPF map.

DamianSawicki replied:
Thanks for the explanation, removed `val` as it's no longer needed.

joamaki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -718,6 +869,70 @@ func (f *FakeLBMaps) UpdateSourceRange(key lbmap.SourceRangeKey, value *lbmap.So
+		outer = &f.mglv4
+	}
+	var singletonKey lbmap.MaglevInnerKey
+	// value, ok := outer.Map.Load(bpfKey(&key))
```
Delete dead code?

DamianSawicki replied:
Yeah, I was keeping this in case we decide to change the logic of `UpdateMaglev`, but we're preserving it for now, so I've deleted this.

joamaki commented on pkg/loadbalancer/experimental/test_utils.go:
```diff
@@ -300,7 +301,31 @@ func DumpLBMaps(lbmaps LBMaps, feAddr loadbalancer.L3n4Addr, sanitizeIDs bool, c
+		for _, id := range trunc {
+			ids = append(ids, sanitizeID(id, sanitizeIDs))
+		}
+		maglev = append(maglev, fmt.Sprintf("MAGLEV: ID=%s INNER=[%s%s]",
```
Hmm, I wonder if there's a better way to visualize this as the current form isn't that useful. E.g. if we assume the maglev table computation is correct then what we really only care about is which ids are mentioned in the table. So we could just show `MAGLEV: ID=123 INNER=[1(10), 2(11), 3(10)]` (e.g. backend `1` appears 10 times in the table).

DamianSawicki replied:
That's a nice idea! But in tests we typically sanitize IDs to `zero` or `non-zero` (see `/testdata`), so we probably also should not rely on their order `1, 2, 3`, right? Perhaps sorting by counts rather than by IDs is the way to go?

DamianSawicki replied:
> Perhaps sorting by counts rather than by IDs is the way to go?

Done.

joamaki commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -866,6 +900,31 @@ func (ops *BPFOps) releaseBackend(id loadbalancer.BackendID, addr loadbalancer.L
 }
 
+func (ops *BPFOps) computeMaglevTable(svc *Service, bes []*BackendWithRevision) ([]loadbalancer.BackendID, error) {
+	backendsMap := make(map[string]*loadbalancer.Backend, len(bes))
```
I would eventually want to get rid of this useless alloc&copy, but I think for now it's fine to keep `maglev.GetLookupTable` as is and then later revisit to clean this up.

DamianSawicki replied:
Yeah, makes sense.

joamaki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -443,6 +486,73 @@ func (r *BPFLBMaps) UpdateSourceRange(key lbmap.SourceRangeKey, value *lbmap.Sou
 
+// UpdateMaglev implements lbmaps.
+func (r *BPFLBMaps) UpdateMaglev(key lbmap.MaglevOuterKey, backendIDs []loadbalancer.BackendID, ipv6 bool) error {
+	// TODO (optional): Consider benchmarking if checking for existence and overwriting an existing map is faster than re-creation.
```
Remove TODO comment

DamianSawicki replied:
Done.

DamianSawicki commented on pkg/maglev/maglev.go:
```diff
@@ -74,6 +85,10 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
+	// As the permutations array is large and often used, we'll keep a single
+	// instance around and reuse it.
+	permutation := permutations.get()
 	if size := uint64(bCount) * m; size > uint64(len(permutation)) {
```
This is going to reallocate for each increment of `bCount`. The previous solution never reallocated in practice and in theory it would only reallocate on the increase of `m`.

joamaki replied:
Fixed.

DamianSawicki commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -437,6 +447,33 @@ func (ops *BPFOps) pruneSourceRanges() error {
+		ipv6 bool
+	}
+	toDelete := []outerKeyWithIPVersion{}
+	cb := func(key lbmap.MaglevOuterKey, value lbmap.MaglevOuterVal, _ lbmap.MaglevInnerKey, _ *lbmap.MaglevInnerVal, ipv6 bool) {
```
Nit: `value` can be replaced with `_`.

DamianSawicki replied:
Fixed.

DamianSawicki commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -882,6 +945,36 @@ func (ops *BPFOps) releaseBackend(id loadbalancer.BackendID, addr loadbalancer.L
+	beSeq := func(yield func(maglev.Backend) bool) {
+		for _, be := range bes {
+			id, err := ops.backendIDAlloc.lookupLocalID(be.L3n4Addr)
+			if err != nil {
```
Are `err != nil` and `!ok` expected here? Shouldn't we log these?

DamianSawicki replied:
I guess it's outdated. These errors are bubbled up and then logs in the current version.

DamianSawicki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -608,6 +649,72 @@ func (r *BPFLBMaps) UpdateSourceRange(key lbmap.SourceRangeKey, value *lbmap.Sou
+		var singletonKey lbmap.MaglevInnerKey
+		innerValue, err := inner.Lookup(&singletonKey)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("lookup up first inner map key (backends) for map with id %d: %w", maglevValue.FD, err))
```
`lookup up` sounds incorrect, and in fact the whole sentence is difficult to parse. I'd suggest something like
`cannot look up backends in inner map with id %d: %w`.

DamianSawicki replied:
Fixed.

DamianSawicki commented on pkg/maglev/maglev.go:
```diff
@@ -7,77 +7,135 @@ import (
+		seedMurmur:   seedMurmur,
+		SeedJhash0:   SeedJhash0,
+		SeedJhash1:   SeedJhash1,
+		backendInfos: make([]backendInfo, 0, 1024),
```
1024 is arbitrary and small. I'd suggest replacing `make([]backendInfo, 0, 1024),` with just `nil`.

joamaki replied:
I'm a bit puzzled why you're worried about reallocations of the permutation slice but you're not worried about this? The permutation slice preallocation is also quite arbitrary.

DamianSawicki replied:
Unless I'm messing something up, the size of the permutation slice is at least `n*n*100`, where `n` is the size of `backendInfos`, so the permutation slice is much bigger.

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

DamianSawicki commented on pkg/maglev/maglev.go:
```diff
@@ -7,77 +7,135 @@ import (
-	// Allocate this ahead of time to avoid expensive allocations inside
-	// getPermutation().
-	permutation = make([]uint64, derivePermutationSliceLen(m))
+	minSize := derivePermutationSliceLen(uint64(cfg.MaglevTableSize))
```
Why do we want to call this size `min`? I'd use just `permutationSize` and perhaps retain the deleted comment about allocation ahead of time (although there's already a comment in the definition of the `Maglev struct`, so it's not strictly necessary).

DamianSawicki commented on pkg/maglev/maglev.go:
```diff
@@ -7,77 +7,135 @@ import (
 		// Reallocate slice so we don't have to allocate again on the next
 		// call.
-		permutation = make([]uint64, size)
+		ml.permutations = make([]uint64, size)
```
Is this code supposed to be a failover mechanism when `bCount` exceeds `m/100`?

If so, it should be marked as such in comments.

If not, I'd add a check somewhere that `bCount * 100 <= m` and replace `make([]uint64, size)` with `make([]uint64, derivePermutationSliceLen(m))` (to avoid reallocation for each increment of `bCount`). Do we have a mechanism in place to update `ValueSize` in `maglevInnerMapSpec` and `InnerMap` (inside `maglev4MapSpec` and `maglev6MapSpec`) when `m` is increased?

joamaki replied:
`m` cannot be increased at runtime (and probably should be consistent across all nodes). This is just checking if the previous permutation array can hold `m*len(backends)` and if not reallocs it. The initial `derivePermutationsSliceLen` is likely large enough already that it's very unlikely we'll perform many allocations. Plus we really don't need to worry about couple reallocations here and there (provided it amortizes to no allocations over time) considering how expensive this computation is.

DamianSawicki replied:
If `m` cannot be increased at runtime, the only situation when the reallocation beyond the initial size is needed is when `bCount` exceeds `m/100`. Could we add a warning log here to indicate it?

The reason why I'm so concerned about this are the exemplar values in the comments above `derivePermutationsSliceLen` saying that the permutation can grow up to 1GB+ (for non-default `m`). I'm not sure if the computation really remains feasible at such scales, but even a single reallocation would cause a memory usage spike by 1GB (until the previous slice is garbage collected), which may be problematic with tight memory limits applied by k8s users to control costs.

Btw. since `m == uint64(ml.Config.MaglevTableSize)`, perhaps we can remove it from the signature of `getPermutation`?

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

DamianSawicki commented on pkg/maglev/maglev.go:
```diff
@@ -89,90 +147,123 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
+		}
+		ml.backendInfos = append(ml.backendInfos, info)
+	}
+	// Sort by address & state
```
How is it supposed to be sorting by state if state is not even a part of the `Backend interface`? Are we assuming that `backends iter.Seq[BE]` is already sorted? If so, why are we sorting again?

joamaki replied:
Oops, I dropped the state and forgot to update the comment.
We can't assume `iter.Seq` is sorted here (it's not for the legacy case)

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

DamianSawicki commented on pkg/maglev/maglev.go:
```diff
@@ -89,90 +147,123 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
+		ml.backendInfos = append(ml.backendInfos, info)
+	}
+	// Sort by address & state
+	slices.SortStableFunc(
```
I believe it's impossible to have two backends with equal addresses here, right? If so, then perhaps using `SortFunc` instead of `SortStableFunc` is sufficient and potentially slightly faster (since we're already optimizing tiny things like reusing `ml.backendInfos`).

joamaki replied:
Sure!

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

DamianSawicki commented on pkg/maglev/maglev.go:
```diff
@@ -89,90 +147,123 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
 // Maglev algorithm might produce different lookup table for the same
 // set of backends listed in a different order. To avoid that sort
-// backends by name, as the names are the same on all nodes (in opposite
+// backends by the hash, as these are the same on all nodes (in opposite
```
It seems to me that `getBackendInfos` sorts by address rather than hash.

joamaki replied:
Oopsie another artifact from previous version. Good catch!

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

DamianSawicki commented on pkg/maglev/maglev.go:
```diff
@@ -89,90 +147,123 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
-	perm := getPermutation(backends, m, runtime.NumCPU())
 	next := make([]int, len(backends))
-	entry := make([]int, m)
+	entry := make([]loadbalancer.BackendID, m)
```
Could you please add a comment that we're not reusing this because it is the return value? At first glance, it looked inconsistent to me that we're reusing `ml.backendInfos` which is 100x shorter and not this one.

joamaki replied:
Good idea. I also had the same confusion :)

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

DamianSawicki commented on pkg/maglev/maglev_test.go:
```diff
@@ -4,33 +4,31 @@
 		perm := make([]uint64, len(backends)*int(m))
 		for i, backend := range backends {
-			offset, skip := getOffsetAndSkip(backend, m)
+			h1, h2 := murmur3.Hash128(backend.addrBytes, ml.seedMurmur)
```
Why is `getOffsetAndSkip` inlined? The inlined code appears here in `pkg/maglev/maglev_test.go` and in `pkg/maglev/maglev.go` so it makes sense to keep a function for it.

DamianSawicki replied:
Outdated, resolving.

DamianSawicki commented on pkg/maglev/maglev_test.go:
```diff
@@ -39,43 +37,66 @@ func TestPermutations(t *testing.T) {
+		if (backend == 1 || backend == 2) && after[pos] != before[pos] {
 			changesInExistingBackends++
 		} else {
 			// Check that "three" placement was overridden by "one" or "two"
```
This comment line should be removed or updated with the removal of `backendsMap := map[string]*loadbalancer.Backend`.

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

DamianSawicki commented on pkg/maglev/maglev_test.go:
```diff
@@ -39,43 +37,66 @@ func TestPermutations(t *testing.T) {
-		"one":   {Weight: 1, ID: 0},
-		"three": {Weight: 1, ID: 1},
-		"two":   {Weight: 1, ID: 2},
+	backends := []*loadbalancer.Backend{
```
Why do we start counting from 1 instead of 0? It leads to changes below that seem unnecessary.

joamaki replied:
I don't like using `0` (e.g. zero values) in test data. It leads to confusion. So bit of a boy scout change. Happy to revert it if you'd prefer (https://registerspill.thorstenball.com/p/use-data-that-looks-like-data sums up my thinking pretty well)

DamianSawicki replied:
Fair enough.

DamianSawicki commented on pkg/maglev/maglev_test.go:
```diff
@@ -85,37 +106,39 @@ func TestBackendRemoval(t *testing.T) {
-		"three": {Weight: 13, ID: 1},  // approx. 97.5x times
-		"two":   {Weight: 111, ID: 2}, // approx. 833x times
-		"tzwe":  {Weight: 10, ID: 3},  // approx. 75x times
+	backends := []*loadbalancer.Backend{
```
Same here. Why do we start counting from 1 instead of 0? It leads to changes below that seem unnecessary.

DamianSawicki replied:
Resolving as above.

DamianSawicki commented on pkg/maglev/maglev_test.go:
```diff
@@ -126,28 +149,38 @@ func TestWeightedBackendWithRemoval(t *testing.T) {
-		backends[fmt.Sprintf("backend-%d", i)] = &loadbalancer.Backend{Weight: 1}
+		backends[i] = &loadbalancer.Backend{Weight: 1, State: loadbalancer.BackendStateActive, L3n4Addr: mkAddr(int32(i))}
 	}
+	infos := getBackendInfos(ml, slices.Values(backends))
```
Previously, sorting backends was a part of GetMaglevTable, so perhaps we should move this part inside the `ResetTimer() - StopTimer()`? Or even better benchmark `GetLookupTable[BE Backend](ml *Maglev, backends iter.Seq[BE]) []loadbalancer.BackendID`, which takes locking into account.

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

DamianSawicki reviewed (commented):
Strange to review my own PR, but it ended up being a collaborative one :D

derailed commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -437,6 +447,33 @@ func (ops *BPFOps) pruneSourceRanges() error {
+	toDelete := []outerKeyWithIPVersion{}
+	cb := func(key lbmap.MaglevOuterKey, value lbmap.MaglevOuterVal, _ lbmap.MaglevInnerKey, _ *lbmap.MaglevInnerVal, ipv6 bool) {
+		if _, ok := ops.serviceIDAlloc.entitiesID[loadbalancer.ID(key.RevNatID)]; !ok {
+			ops.log.Debug("pruneMaglev: deleting", "id", key.RevNatID)
```
nit: we are not really deleting here.

DamianSawicki replied:
It was copy-pasted from other `prune...` functions in this file, but you're right, I'll fix all of them.

derailed commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -437,6 +447,33 @@ func (ops *BPFOps) pruneSourceRanges() error {
+	}
+	for _, okwiv := range toDelete {
+		err := ops.LBMaps.DeleteMaglev(okwiv.MaglevOuterKey, okwiv.ipv6)
+		if err != nil {
```
Should we let the call site determine what to do if deletion errs have occurred vs logging?

DamianSawicki replied:
Again, it is the same as in other analogous places in this file, but indeed it's kind of strange (@joamaki?) that `Prune()` joins all the errors from `pruneXXX()`, but these functions sometimes don't bubble up their errors. I'll change `pruneMaglev()` and see if it doesn't break anything before proceeding to other functions.

derailed commented on pkg/loadbalancer/experimental/bpf_reconciler.go:
```diff
@@ -811,6 +857,23 @@ func (ops *BPFOps) upsertRevNat(id loadbalancer.ID, svcKey lbmap.ServiceKey, svc
 }
 
+func (ops *BPFOps) updateMaglev(fe *Frontend, feID loadbalancer.ID, activeBackends []BackendWithRevision) error {
+	if len(activeBackends) > 0 {
```
nit: reverse logic to eliminate the indents i.e if len(activeBackends) == 0 { return xxx}

DamianSawicki replied:
Reversing the logic doesn't suffice, but at the price of an additional `return` I can indeed eliminate the indents. Will do.

derailed commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -466,6 +466,15 @@ func (b *Backend) String() string {
 
+// WeightInfo implements interface needed for maglev to compute
+// the maglev lookup tables for backends.
+func (b *Backend) WeightInfo() (id BackendID, addr L3n4Addr, weight uint16) {
```
nit: named returns here seems superfluous.

DamianSawicki replied:
FYI: This part of the PR will most likely be moved to a separate PR.

This was authored by @joamaki's, but I think named arguments are useful here for documentation purposes because neither the function name nor the doc-string specify what the function computes. One can probably infer that from the types, although `uint16` is not very descriptive...

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

derailed commented on pkg/maglev/maglev.go:
```diff
@@ -89,90 +147,123 @@ func getPermutation(backends []string, m uint64, numCPU int) []uint64 {
+
+func getBackendInfos[BE Backend](ml *Maglev, backends iter.Seq[BE]) []backendInfo {
+	ml.backendInfos = ml.backendInfos[:0]
+	for b := range backends {
```
This is a little confusing. The interface Backend returns the entire object state at least at this juncture.  Thus why not just export and return a BackendInfo?
Also the func name is misleading since we are actually `re-setting` backends and returning them.

DamianSawicki replied:
FYI: This part of the PR will most likely be moved to a separate PR.

Again, this was authored by @joamaki, but let me share my perspective.

> Also the func name is misleading since we are actually re-setting backends and returning them.

`ml.backendInfos` is an un-exported field of type `Maglev` and it is clearly marked as a reusable buffer in the type definition, so I think it's fine to modify it "silently". It would be more natural to do it in a method than in a standalone function, but I guess it is impossible with generics, right?

> The interface Backend returns the entire object state at least at this juncture. Thus why not just export and return a BackendInfo?

I believe the point is to use `GetLookupTable[Be Backend]` with different actual types `BE` implementing the `maglev.Backend` interface. It is used with `experimental.maglevBackend` in `pkg/loadbalancer/experimental/bpf_reconciler.go` and with `*loadbalancer.Backend` in `pkg/maps/lbmap/lbmap.go`.

joamaki replied:
I think I'll refactor this code to have `type BackendInfo struct` and move `GetLookupTable` back into a method and take `iter.Seq[BackendInfo]`. This also allows us to avoid the allocations and temporary data structures, but it's much more direct.

DamianSawicki replied:
Moved to https://github.com/cilium/cilium/pull/35885, resolving.

derailed reviewed (commented):
@DamianSawicki Nice work!

DamianSawicki committed:
experimental: Add Maglev support

This adds support for the Maglev load-balancing algorithm to the
experimental StateDB load balancing.

Signed-off-by: Damian Sawicki <dsawicki@google.com>
Signed-off-by: Jussi Maki <jussi@isovalent.com>

jrajahalme reviewed (approved):
Envoy related change seems OK :-)
